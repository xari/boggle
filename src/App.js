import React, { useEffect, useState } from "react";
import { useAsyncMemo } from "use-async-memo";
import Controls from "./Controls";
import Board from "./Board";
import { solve, createRandomBoard } from "./utils";
import "./App.css";

function Results({ board }) {
  // Load and memoize the dictionary
  const dictionary = useAsyncMemo(
    async () =>
      await import("./dictionary_en_US.json").then((data) =>
        data.words.filter((word) => word.length >= 3)
      ),
    []
  );

  const [results, setResults] = useState(null);

  useEffect(() => {
    dictionary && setResults(board !== null ? solve(board, dictionary) : null);
  }, [board, dictionary]);

  const { words, timeToSolve } = results !== null && results;

  return (
    <div className="my-5">
      {typeof timeToSolve !== "undefined" && (
        <p className="text-lg">
          It took{" "}
          <span className="font-semibold">
            {timeToSolve.toFixed(2)} milliseconds
          </span>{" "}
          to solve this board
          <br />
          All possible words on the board are shown below.
        </p>
      )}
      <div className="flex flex-wrap py-2 -mx-1 w-full">
        {typeof words !== "undefined" &&
          Array.from(words).map((word, i) => (
            <div
              key={i}
              className="m-1 px-2 py-0 text-sm font-light text-emerald-800 bg-gray-100 border border-emerald-800 rounded-lg"
            >
              {word}
            </div>
          ))}
      </div>
    </div>
  );
}

function App() {
  // State
  const [dimensions, setDimensions] = useState(4);
  const [enableRandom, setEnableRandom] = useState(true);
  const [board, setBoard] = useState(createRandomBoard(dimensions));

  // Controlled state updaters
  const updateDimensions = (e) => setDimensions(parseInt(e.target.value));
  const updateEnableRandom = () => setEnableRandom(!enableRandom);
  const updateBoard = (newBoard) => {
    setBoard(newBoard);
  };

  return (
    <div className="my-8 mx-auto w-full max-w-xl px-4 sm:px-6 lg:px-8">
      <h1 className="mt-5 mb-8 text-8xl text-center text-emerald-500">
        Boggle!
      </h1>
      <Controls
        dimensions={dimensions}
        setDimensions={updateDimensions}
        enableRandom={enableRandom}
        setEnableRandom={updateEnableRandom}
      />
      <Board
        board={board}
        setBoard={updateBoard}
        dimensions={dimensions}
        enableRandom={enableRandom}
      />
      <Results board={board} />
    </div>
  );
}

export default App;
