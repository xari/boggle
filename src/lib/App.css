.font-stroke {
  -webkit-text-stroke: 2px black;
}

.font-typewriter {
  font-family: American Typewriter;
}
