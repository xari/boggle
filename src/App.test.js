import React from "react";
import renderer from "react-test-renderer";
import { Results } from "./App";

describe("the Results", () => {
  const board = [
    ["v", "z", "g", "x"],
    ["j", "z", "g", "b"],
    ["t", "b", "j", "r"],
    ["u", "a", "c", "p"],
  ];

  it("render without any unexpected changes", () => {
    const tree = renderer.create(<Results board={board} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
