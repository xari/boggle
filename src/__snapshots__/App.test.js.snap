// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Results render without any unexpected changes 1`] = `
<div
  className="my-5"
>
  <div
    className="flex flex-wrap py-2 -mx-1 w-full"
  />
</div>
`;
